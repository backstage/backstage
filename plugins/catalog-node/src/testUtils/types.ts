/*
 * Copyright 2024 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  AddLocationRequest,
  AddLocationResponse,
  CatalogApi,
  CatalogRequestOptions,
  GetEntitiesByRefsRequest,
  GetEntitiesByRefsResponse,
  GetEntitiesRequest,
  GetEntitiesResponse,
  GetEntityAncestorsRequest,
  GetEntityAncestorsResponse,
  GetEntityFacetsRequest,
  GetEntityFacetsResponse,
  Location,
  QueryEntitiesRequest,
  QueryEntitiesResponse,
  ValidateEntityResponse,
} from '@backstage/catalog-client';
import { CompoundEntityRef, Entity } from '@backstage/catalog-model';
import {
  CatalogService,
  CatalogServiceRequestOptions,
} from '@backstage/plugin-catalog-node';

/**
 * A mock friendly version of `CatalogService` /
 * {@link @backstage/catalog-client#CatalogApi | CatalogApi}.
 *
 * @public
 * @remarks
 *
 * This interface supports both API types at the same time, and has an optional
 * second argument to simplify testing since the mock implementation does not
 * care about credentials.
 */
export interface CatalogServiceMock extends CatalogService, CatalogApi {
  getEntities(
    request?: GetEntitiesRequest,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<GetEntitiesResponse>;

  getEntitiesByRefs(
    request: GetEntitiesByRefsRequest,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<GetEntitiesByRefsResponse>;

  queryEntities(
    request?: QueryEntitiesRequest,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<QueryEntitiesResponse>;

  getEntityAncestors(
    request: GetEntityAncestorsRequest,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<GetEntityAncestorsResponse>;

  getEntityByRef(
    entityRef: string | CompoundEntityRef,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<Entity | undefined>;

  removeEntityByUid(
    uid: string,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<void>;

  refreshEntity(
    entityRef: string,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<void>;

  getEntityFacets(
    request: GetEntityFacetsRequest,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<GetEntityFacetsResponse>;

  getLocationById(
    id: string,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<Location | undefined>;

  getLocationByRef(
    locationRef: string,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<Location | undefined>;

  addLocation(
    location: AddLocationRequest,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<AddLocationResponse>;

  removeLocationById(
    id: string,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<void>;

  getLocationByEntity(
    entityRef: string | CompoundEntityRef,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<Location | undefined>;

  validateEntity(
    entity: Entity,
    locationRef: string,
    options?: CatalogServiceRequestOptions | CatalogRequestOptions,
  ): Promise<ValidateEntityResponse>;
}
