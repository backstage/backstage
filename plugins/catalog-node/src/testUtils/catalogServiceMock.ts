/*
 * Copyright 2024 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ServiceRef,
  createServiceFactory,
} from '@backstage/backend-plugin-api';
import { InMemoryCatalogClient } from '@backstage/catalog-client/testUtils';
import { Entity } from '@backstage/catalog-model';
import { catalogServiceRef } from '@backstage/plugin-catalog-node/alpha';
// eslint-disable-next-line @backstage/no-undeclared-imports
import { ServiceMock } from '@backstage/backend-test-utils';
import { CatalogApi } from '@backstage/catalog-client';

/** @internal */
function simpleMock<TService>(
  ref: ServiceRef<TService, any>,
  mockFactory: () => jest.Mocked<TService>,
): (partialImpl?: Partial<TService>) => ServiceMock<TService> {
  return partialImpl => {
    const mock = mockFactory();
    if (partialImpl) {
      for (const [key, impl] of Object.entries(partialImpl)) {
        if (typeof impl === 'function') {
          (mock as any)[key].mockImplementation(impl);
        } else {
          (mock as any)[key] = impl;
        }
      }
    }
    return Object.assign(mock, {
      factory: createServiceFactory({
        service: ref,
        deps: {},
        factory: () => mock,
      }),
    }) as ServiceMock<TService>;
  };
}

/**
 * Creates a fake catalog client that handles entities in memory storage. Note
 * that this client may be severely limited in functionality, and advanced
 * functions may not be available at all.
 *
 * @public
 */
export function catalogServiceMock(options?: {
  entities?: Entity[];
}): CatalogApi {
  return new InMemoryCatalogClient(options);
}

/**
 * A collection of mock functionality for the catalog service.
 *
 * @public
 */
export namespace catalogServiceMock {
  /**
   * Creates a fake catalog client that handles entities in memory storage. Note
   * that this client may be severely limited in functionality, and advanced
   * functions may not be available at all.
   */
  export const factory = (options?: { entities?: Entity[] }) =>
    createServiceFactory({
      service: catalogServiceRef,
      deps: {},
      factory: () => new InMemoryCatalogClient(options),
    });
  /**
   * Creates a catalog client whose methods are mock functions, possibly with
   * some of them overloaded by the caller.
   */
  export const mock = simpleMock(catalogServiceRef, () => ({
    getEntities: jest.fn(),
    getEntitiesByRefs: jest.fn(),
    queryEntities: jest.fn(),
    getEntityAncestors: jest.fn(),
    getEntityByRef: jest.fn(),
    removeEntityByUid: jest.fn(),
    refreshEntity: jest.fn(),
    getEntityFacets: jest.fn(),
    getLocationById: jest.fn(),
    getLocationByRef: jest.fn(),
    addLocation: jest.fn(),
    removeLocationById: jest.fn(),
    getLocationByEntity: jest.fn(),
    validateEntity: jest.fn(),
  }));
}
