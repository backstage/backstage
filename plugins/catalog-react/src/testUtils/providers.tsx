/*
 * Copyright 2021 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  PropsWithChildren,
  useCallback,
  useMemo,
  useState,
} from 'react';
import {
  DefaultEntityFilters,
  EntityListContext,
  EntityListContextProps,
} from '../hooks/useEntityListProvider';

/** @public */
export function MockEntityListContextProvider<
  T extends DefaultEntityFilters = DefaultEntityFilters,
>(
  props: PropsWithChildren<{
    value?: Partial<EntityListContextProps<T>>;
  }>,
) {
  const { children, value } = props;

  // Provides a default implementation that stores filter state, for testing components that
  // reflect filter state.
  const [filters, setFilters] = useState<T>(value?.filters ?? ({} as T));

  const updateFilters = useCallback(
    (update: Partial<T> | ((prevFilters: T) => Partial<T>)) => {
      setFilters(prevFilters => {
        const newFilters =
          typeof update === 'function' ? update(prevFilters) : update;
        return { ...prevFilters, ...newFilters };
      });
    },
    [],
  );

  // Memoize the default values since pickers have useEffect triggers on these; naively defaulting
  // below with `?? <X>` breaks referential equality on subsequent updates.
  const defaultValues = useMemo(
    () => ({
      entities: [],
      backendEntities: [],
      queryParameters: {},
    }),
    [],
  );

  const resolvedValue: EntityListContextProps<T> = useMemo(
    () => ({
      entities: value?.entities ?? defaultValues.entities,
      backendEntities: value?.backendEntities ?? defaultValues.backendEntities,
      updateFilters: value?.updateFilters ?? updateFilters,
      filters,
      loading: value?.loading ?? false,
      queryParameters: value?.queryParameters ?? defaultValues.queryParameters,
      error: value?.error,
      totalItems:
        value?.totalItems ?? (value?.entities ?? defaultValues.entities).length,
      limit: value?.limit ?? 20,
      offset: value?.offset,
      setLimit: value?.setLimit ?? (() => {}),
      setOffset: value?.setOffset,
      paginationMode: value?.paginationMode ?? 'none',
    }),
    [value, defaultValues, filters, updateFilters],
  );

  return (
    <EntityListContext.Provider value={resolvedValue}>
      {children}
    </EntityListContext.Provider>
  );
}
