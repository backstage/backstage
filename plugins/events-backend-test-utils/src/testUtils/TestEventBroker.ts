/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  EventBroker,
  EventParams,
  EventSubscriber,
} from '@backstage/plugin-events-node';

/**
 * @public
 * @deprecated use `TestEventsService` instead
 */
export class TestEventBroker implements EventBroker {
  readonly published: EventParams[] = [];
  readonly subscribed: EventSubscriber[] = [];

  async publish(params: EventParams): Promise<void> {
    this.published.push(params);
  }

  subscribe(
    ...subscribers: Array<EventSubscriber | Array<EventSubscriber>>
  ): void {
    this.subscribed.push(...subscribers.flat());
  }
}
