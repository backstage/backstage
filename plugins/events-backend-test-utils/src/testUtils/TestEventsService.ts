/*
 * Copyright 2024 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  EventParams,
  EventsService,
  EventsServiceSubscribeOptions,
} from '@backstage/plugin-events-node';

/** @public */
export class TestEventsService implements EventsService {
  #published: EventParams[] = [];
  #subscribed: EventsServiceSubscribeOptions[] = [];

  async publish(params: EventParams): Promise<void> {
    this.#published.push(params);
  }

  async subscribe(options: EventsServiceSubscribeOptions): Promise<void> {
    this.#subscribed.push(options);
  }

  get published(): EventParams[] {
    return this.#published;
  }

  get subscribed(): EventsServiceSubscribeOptions[] {
    return this.#subscribed;
  }

  reset(): void {
    this.#published = [];
    this.#subscribed = [];
  }
}
