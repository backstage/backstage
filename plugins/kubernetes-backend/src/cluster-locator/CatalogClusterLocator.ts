/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  AuthService,
  BackstageCredentials,
} from '@backstage/backend-plugin-api';
import { ClusterDetails, KubernetesClustersSupplier } from '../types/types';
import { CATALOG_FILTER_EXISTS, CatalogApi } from '@backstage/catalog-client';
import {
  ANNOTATION_KUBERNETES_API_SERVER,
  ANNOTATION_KUBERNETES_API_SERVER_CA,
  ANNOTATION_KUBERNETES_AUTH_PROVIDER,
  ANNOTATION_KUBERNETES_SKIP_METRICS_LOOKUP,
  ANNOTATION_KUBERNETES_SKIP_TLS_VERIFY,
  ANNOTATION_KUBERNETES_DASHBOARD_URL,
  ANNOTATION_KUBERNETES_DASHBOARD_APP,
  ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS,
} from '@backstage/plugin-kubernetes-common';
import { JsonObject } from '@backstage/types';

function isObject(obj: unknown): obj is JsonObject {
  return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}

export class CatalogClusterLocator implements KubernetesClustersSupplier {
  private catalogClient: CatalogApi;
  private auth: AuthService;

  constructor(catalogClient: CatalogApi, auth: AuthService) {
    this.catalogClient = catalogClient;
    this.auth = auth;
  }

  static fromConfig(
    catalogApi: CatalogApi,
    auth: AuthService,
  ): CatalogClusterLocator {
    return new CatalogClusterLocator(catalogApi, auth);
  }

  async getClusters(options?: {
    credentials: BackstageCredentials;
  }): Promise<ClusterDetails[]> {
    const apiServerKey = `metadata.annotations.${ANNOTATION_KUBERNETES_API_SERVER}`;
    const apiServerCaKey = `metadata.annotations.${ANNOTATION_KUBERNETES_API_SERVER_CA}`;
    const authProviderKey = `metadata.annotations.${ANNOTATION_KUBERNETES_AUTH_PROVIDER}`;

    const filter: Record<string, symbol | string> = {
      kind: 'Resource',
      'spec.type': 'kubernetes-cluster',
      [apiServerKey]: CATALOG_FILTER_EXISTS,
      [apiServerCaKey]: CATALOG_FILTER_EXISTS,
      [authProviderKey]: CATALOG_FILTER_EXISTS,
    };

    const clusters = await this.catalogClient.getEntities(
      {
        filter: [filter],
      },
      options?.credentials
        ? {
            token: (
              await this.auth.getPluginRequestToken({
                onBehalfOf: options.credentials,
                targetPluginId: 'catalog',
              })
            ).token,
          }
        : undefined,
    );
    return clusters.items.map(entity => {
      const annotations = entity.metadata.annotations!;
      const clusterDetails: ClusterDetails = {
        name: entity.metadata.name,
        title: entity.metadata.title,
        url: annotations[ANNOTATION_KUBERNETES_API_SERVER],
        authMetadata: annotations,
        caData: annotations[ANNOTATION_KUBERNETES_API_SERVER_CA],
        skipMetricsLookup:
          annotations[ANNOTATION_KUBERNETES_SKIP_METRICS_LOOKUP] === 'true',
        skipTLSVerify:
          annotations[ANNOTATION_KUBERNETES_SKIP_TLS_VERIFY] === 'true',
        dashboardUrl: annotations[ANNOTATION_KUBERNETES_DASHBOARD_URL],
        dashboardApp: annotations[ANNOTATION_KUBERNETES_DASHBOARD_APP],
        dashboardParameters: this.getDashboardParameters(annotations),
      };

      return clusterDetails;
    });
  }

  private getDashboardParameters(
    annotations: Record<string, string>,
  ): JsonObject | undefined {
    const dashboardParamsString =
      annotations[ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS];
    if (dashboardParamsString) {
      try {
        const dashboardParams = JSON.parse(dashboardParamsString);
        return isObject(dashboardParams) ? dashboardParams : undefined;
      } catch {
        return undefined;
      }
    }
    return undefined;
  }
}
