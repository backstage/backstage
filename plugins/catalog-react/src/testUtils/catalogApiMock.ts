/*
 * Copyright 2024 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ApiFactory,
  ApiRef,
  createApiFactory,
} from '@backstage/frontend-plugin-api';
import { InMemoryCatalogClient } from '@backstage/catalog-client/testUtils';
import { Entity } from '@backstage/catalog-model';
import { catalogApiRef } from '@backstage/plugin-catalog-react';
import { CatalogApi } from '@backstage/catalog-client';
import { ApiMock } from '@backstage/frontend-test-utils';

/** @internal */
function simpleMock<TApi>(
  ref: ApiRef<TApi>,
  mockFactory: () => jest.Mocked<TApi>,
): (partialImpl?: Partial<TApi>) => ApiMock<TApi> {
  return partialImpl => {
    const mock = mockFactory();
    if (partialImpl) {
      for (const [key, impl] of Object.entries(partialImpl)) {
        if (typeof impl === 'function') {
          (mock as any)[key].mockImplementation(impl);
        } else {
          (mock as any)[key] = impl;
        }
      }
    }
    return Object.assign(mock, {
      factory: createApiFactory({
        api: ref,
        deps: {},
        factory: () => mock,
      }),
    }) as ApiMock<TApi>;
  };
}

/**
 * Creates a fake catalog client that handles entities in memory storage. Note
 * that this client may be severely limited in functionality, and advanced
 * functions may not be available at all.
 *
 * @public
 */
export function catalogApiMock(options?: { entities?: Entity[] }): CatalogApi {
  return new InMemoryCatalogClient(options);
}

/**
 * A collection of mock functionality for the catalog service.
 *
 * @public
 */
export namespace catalogApiMock {
  /**
   * Creates a fake catalog client that handles entities in memory storage. Note
   * that this client may be severely limited in functionality, and advanced
   * functions may not be available at all.
   */
  export const factory = (options?: {
    entities?: Entity[];
  }): ApiFactory<CatalogApi, CatalogApi, {}> =>
    createApiFactory({
      api: catalogApiRef,
      deps: {},
      factory: () => new InMemoryCatalogClient(options),
    });
  /**
   * Creates a catalog client whose methods are mock functions, possibly with
   * some of them overloaded by the caller.
   */
  export const mock = simpleMock(catalogApiRef, () => ({
    getEntities: jest.fn(),
    getEntitiesByRefs: jest.fn(),
    queryEntities: jest.fn(),
    getEntityAncestors: jest.fn(),
    getEntityByRef: jest.fn(),
    removeEntityByUid: jest.fn(),
    refreshEntity: jest.fn(),
    getEntityFacets: jest.fn(),
    getLocationById: jest.fn(),
    getLocationByRef: jest.fn(),
    addLocation: jest.fn(),
    removeLocationById: jest.fn(),
    getLocationByEntity: jest.fn(),
    validateEntity: jest.fn(),
  }));
}
