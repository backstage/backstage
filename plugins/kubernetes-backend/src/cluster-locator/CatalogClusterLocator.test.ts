/*
 * Copyright 2020 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '@backstage/backend-common';
import {
  ANNOTATION_KUBERNETES_AUTH_PROVIDER,
  ANNOTATION_KUBERNETES_AWS_ASSUME_ROLE,
  ANNOTATION_KUBERNETES_AWS_EXTERNAL_ID,
  ANNOTATION_KUBERNETES_OIDC_TOKEN_PROVIDER,
} from '@backstage/plugin-kubernetes-common';
import { CatalogClusterLocator } from './CatalogClusterLocator';
import { mockCredentials, mockServices } from '@backstage/backend-test-utils';
import { Entity } from '@backstage/catalog-model';
import { catalogServiceMock } from '@backstage/plugin-catalog-node/testUtils';

const entities: Entity[] = [
  {
    apiVersion: 'backstage.io/v1alpha1',
    kind: 'Resource',
    metadata: {
      annotations: {
        'kubernetes.io/api-server': 'https://apiserver.com',
        'kubernetes.io/api-server-certificate-authority': 'caData',
        [ANNOTATION_KUBERNETES_AUTH_PROVIDER]: 'oidc',
        [ANNOTATION_KUBERNETES_OIDC_TOKEN_PROVIDER]: 'google',
        'kubernetes.io/skip-metrics-lookup': 'true',
        'kubernetes.io/skip-tls-verify': 'true',
        'kubernetes.io/dashboard-url': 'my-url',
        'kubernetes.io/dashboard-app': 'my-app',
      },
      name: 'owned',
      title: 'title',
      namespace: 'default',
    },
    spec: {
      type: 'kubernetes-cluster',
    },
  },
  {
    apiVersion: 'backstage.io/v1alpha1',
    kind: 'Resource',
    metadata: {
      annotations: {
        'kubernetes.io/api-server': 'https://apiserver.com',
        'kubernetes.io/api-server-certificate-authority': 'caData',
        [ANNOTATION_KUBERNETES_AUTH_PROVIDER]: 'aws',
        [ANNOTATION_KUBERNETES_AWS_ASSUME_ROLE]: 'my-role',
        [ANNOTATION_KUBERNETES_AWS_EXTERNAL_ID]: 'my-id',
        [ANNOTATION_KUBERNETES_OIDC_TOKEN_PROVIDER]: 'google',
        'kubernetes.io/dashboard-url': 'my-url',
        'kubernetes.io/dashboard-app': 'my-app',
      },
      name: 'owned',
      namespace: 'default',
    },
    spec: {
      type: 'kubernetes-cluster',
    },
  },
];

describe('CatalogClusterLocator', () => {
  it('returns empty cluster details when the cluster is empty', async () => {
    const credentials = mockCredentials.user();
    const clusterSupplier = CatalogClusterLocator.fromConfig(
      catalogServiceMock({ entities: [] }),
      mockServices.auth(),
    );

    const result = await clusterSupplier.getClusters({ credentials });
    expect(result).toHaveLength(0);
    expect(result).toStrictEqual([]);
  });

  it('returns the cluster details provided by annotations', async () => {
    const credentials = mockCredentials.user();
    const clusterSupplier = CatalogClusterLocator.fromConfig(
      catalogServiceMock({ entities }),
      mockServices.auth(),
    );

    const result = await clusterSupplier.getClusters({ credentials });
    expect(result).toHaveLength(2);
    expect(result[0]).toMatchSnapshot();
  });

  it('returns the aws cluster details provided by annotations', async () => {
    const credentials = mockCredentials.user();
    const clusterSupplier = CatalogClusterLocator.fromConfig(
      catalogServiceMock({ entities }),
      mockServices.auth(),
    );

    const result = await clusterSupplier.getClusters({ credentials });
    expect(result).toHaveLength(2);
    expect(result[1]).toMatchSnapshot();
  });
});
