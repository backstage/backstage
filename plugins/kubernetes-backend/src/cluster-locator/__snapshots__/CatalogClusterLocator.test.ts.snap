// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogClusterLocator returns the aws cluster details provided by annotations 1`] = `
{
  "authMetadata": {
    "kubernetes.io/api-server": "https://apiserver.com",
    "kubernetes.io/api-server-certificate-authority": "caData",
    "kubernetes.io/auth-provider": "aws",
    "kubernetes.io/aws-assume-role": "my-role",
    "kubernetes.io/aws-external-id": "my-id",
    "kubernetes.io/dashboard-app": "my-app",
    "kubernetes.io/dashboard-url": "my-url",
    "kubernetes.io/oidc-token-provider": "google",
  },
  "caData": "caData",
  "dashboardApp": "my-app",
  "dashboardParameters": undefined,
  "dashboardUrl": "my-url",
  "name": "owned",
  "skipMetricsLookup": false,
  "skipTLSVerify": false,
  "title": undefined,
  "url": "https://apiserver.com",
}
`;

exports[`CatalogClusterLocator returns the cluster details provided by annotations 1`] = `
{
  "authMetadata": {
    "kubernetes.io/api-server": "https://apiserver.com",
    "kubernetes.io/api-server-certificate-authority": "caData",
    "kubernetes.io/auth-provider": "oidc",
    "kubernetes.io/dashboard-app": "my-app",
    "kubernetes.io/dashboard-url": "my-url",
    "kubernetes.io/oidc-token-provider": "google",
    "kubernetes.io/skip-metrics-lookup": "true",
    "kubernetes.io/skip-tls-verify": "true",
  },
  "caData": "caData",
  "dashboardApp": "my-app",
  "dashboardParameters": undefined,
  "dashboardUrl": "my-url",
  "name": "owned",
  "skipMetricsLookup": true,
  "skipTLSVerify": true,
  "title": "title",
  "url": "https://apiserver.com",
}
`;
